base/loading.jl

```diff
@@ -280,6 +293,46 @@ function include_dependency(path::AbstractString)
     return nothing
 end
 
+# used to track optional dependencies during precompilation
+const _optional_dependencies = String[]
+
+"""
+    register_optional_dependency(mod::String)
+
+In a module, declare that the module with the name `mod` is a dependency
+for precompilation; that is, the module will need to be recompiled if this
+optional module becomes available in the future.
+
+This is only needed if your module depends on a module that is not used via
+`require`, `using`, `import`, or `importall`. It has no effect outside
+pre-compilation.
+
+Most users will want to use instead `@dependson`
+"""
+function register_optional_dependency(mod::String)
+    if myid() == 1 && _track_dependencies[]
+        push!(_optional_dependencies, mod)
+    end
+end
+
+"""
+    dependson(mod, expr)
+
+Marks a expression to be dependent on the availability of a module by name `mod`.
+This only affects precompilation.
+
+See `register_optional_dependency` for more information.
```

Comment:
If this will be included in the docs (rather than simply REPL docstrings), you could make this a Documenter reference, i.e.
```
See [`register_optional_dependency`](@ref) for more information.
```